src/utils/lazy.ts

```diff
@@ -7,21 +7,68 @@ export class Lazy<T> {
     private _isValueCreated: boolean = false;
     private _value: T | undefined;
 
-    public constructor(private readonly valueFactory: () => T) {
+    public constructor(private readonly valueFactory: () => T, private readonly _valueLifetime?: number) {
     }
 
     public get isValueCreated(): boolean {
         return this._isValueCreated;
     }
 
     public get value(): T {
-        if (!this._isValueCreated) {
-            this._value = this.valueFactory();
-            this._isValueCreated = true;
+        if (this._isValueCreated) {
+            return this._value;
+        }
+
+        this._value = this.valueFactory();
+        this._isValueCreated = true;
+
+        if (this._valueLifetime) {
+            const reset = setTimeout(() => {
+                this._isValueCreated = false;
+                this._value = undefined;
+                clearTimeout(reset);
+            }, this._valueLifetime);
         }
 
         return this._value;
     }
 }
 
-export default Lazy;
+export class AsyncLazy<T> {
+    private _isValueCreated: boolean = false;
+    private _value: T | undefined;
+    private _valuePromise: Promise<T> | undefined;
+
+    public constructor(private readonly valueFactory: () => Promise<T>, private readonly _valueLifetime?: number) {
+    }
+
+    public async getValue(): Promise<T> {
+        if (this._isValueCreated) {
+            return this._value;
+        }
+
+        this._isValueCreated = false;
+
+        const isPrimaryPromise = this._valuePromise === undefined; // The first caller is "primary"
+        const fnLocalPromiseRef = this._valuePromise = this._valuePromise ?? this.valueFactory(); // Await any currently-running Promise if there is one
+        this._value = await fnLocalPromiseRef;
+        this._valuePromise = undefined;
+
+        this._isValueCreated = true;
+
+        if (this._valueLifetime && isPrimaryPromise) {
+            const reset = setTimeout(() => {
+                // Will only clear out values if there isn't a currently-running Promise
+                // If there is, this timer will skip, but when that Promise finishes it will go through this code and register a new timer
+                if (this._valuePromise === undefined) {
+                    this._isValueCreated = false;
+                    this._value = undefined;
+                }
+
+                clearTimeout(reset);
```

Comment:
Will this timer ever started twice? I think not, because of the isPrimaryPromise check, so what is the need to clear the timer?

Reply:
The timers are like disposables, and `clearTimeout` is how you "dispose" of them. This is a necessary step.